File: backend/src/main/java/com/hiragana/service/CharacterService.java
package com.hiragana.service;

import com.hiragana.model.Character;
import com.hiragana.model.CharacterType;
import com.hiragana.model.StrokeResult;
import com.hiragana.repository.CharacterRepository;
import com.hiragana.repository.StrokeResultRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.dao.DataAccessException;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class CharacterService {
    private static final Logger logger = LoggerFactory.getLogger(CharacterService.class);

    @Autowired
    private CharacterRepository characterRepository;

    @Autowired
    private StrokeResultRepository strokeResultRepository;

    @Autowired
    private StrokeResultService strokeResultService;

    public List<Character> getAllCharacters() {
        try {
            logger.debug("Fetching all characters from database");
            List<Character> characters = characterRepository.findAll();
            logger.debug("Found {} characters", characters.size());
            return characters;
        } catch (DataAccessException e) {
            logger.error("Database error while fetching all characters", e);
            throw new RuntimeException("Failed to fetch characters from database", e);
        } catch (Exception e) {
            logger.error("Unexpected error while fetching all characters", e);
            throw new RuntimeException("Unexpected error while fetching characters", e);
        }
    }

    public Optional<Character> getCharacterById(Long id) {
        return characterRepository.findById(id);
    }

    public List<Character> getCharactersByType(CharacterType type) {
        return characterRepository.findByType(type);
    }

    public List<Character> getCharactersByTypeAndDifficulty(CharacterType type, Integer maxDifficulty) {
        return characterRepository.findByTypeAndDifficultyLessThanEqual(type, maxDifficulty);
    }

    public Optional<Character> getCharacterByTypeAndCharacter(CharacterType type, String character) {
        return characterRepository.findByTypeAndCharacter(type, character);
    }

    @Transactional
    public StrokeResult saveStrokeResult(Long characterId, Long userId, StrokeResult strokeResult) {
        return strokeResultService.saveStrokeResult(characterId, userId, strokeResult);
    }

    public Optional<StrokeResult> getStrokeResult(Long characterId, Long userId) {
        return strokeResultService.getStrokeResult(characterId, userId);
    }

    public List<StrokeResult> getAllStrokeResults(Long characterId, Long userId) {
        return strokeResultService.getAllStrokeResults(characterId, userId);
    }

    public StrokeResult findByCharacterIdAndPosition(Long characterId, Integer position, Long userId) {
        return strokeResultRepository.findByCharacter_IdAndPositionAndUser_Id(characterId, position, userId)
            .orElse(null);
    }

    public Character createCharacter(Character character) {
        return characterRepository.save(character);
    }

    public Character updateCharacter(Character character) {
        if (!characterRepository.existsById(character.getId())) {
            throw new RuntimeException("Character not found");
        }
        return characterRepository.save(character);
    }

    public void deleteCharacter(Long id) {
        characterRepository.deleteById(id);
    }

    public void deleteStrokeResult(StrokeResult strokeResult) {
        strokeResultRepository.delete(strokeResult);
    }

    public List<StrokeResult> getAllStrokeResultsByCharacterIdAndUserId(Long characterId, Long userId) {
        return strokeResultRepository.findAllByCharacter_IdAndUser_Id(characterId, userId);
    }
} 

File: frontend/src/api/characterService.js
import apiClient from './client';

export const characterService = {
    // 文字一覧を取得
    async getAllCharacters() {
        const response = await apiClient.get('/api/characters');
        return response.data;
    },

    // 特定のタイプの文字を取得
    getCharactersByType: async (type) => {
        const response = await apiClient.get(`/api/characters/type/${type}`);
        return response.data;
    },

    // 特定のタイプと難易度以下の文字を取得
    getCharactersByTypeAndDifficulty: async (type, maxDifficulty) => {
        const response = await apiClient.get(`/api/characters/type/${type}/difficulty/${maxDifficulty}`);
        return response.data;
    },

    // 特定の文字を取得
    getCharacterByTypeAndCharacter: async (type, character) => {
        const response = await apiClient.get(`/api/characters/type/${type}/character/${character}`);
        return response.data;
    },

    // なぞり結果を保存
    saveStrokeResult: async (characterId, strokeData) => {
        try {
            if (!characterId) {
                throw new Error('文字IDが設定されていません');
            }
            if (strokeData.position === undefined || strokeData.position === null) {
                throw new Error('位置情報が設定されていません');
            }

            console.log('Saving stroke result:', strokeData);
            const strokes = Array.isArray(strokeData.strokes) ? strokeData.strokes : [];
            
            const requestBody = {
                position: strokeData.position,
                strokes: strokes.map(stroke => ({
                    points: Array.isArray(stroke.points) ? stroke.points.map(point => ({
                        x: point.x,
                        y: point.y
                    })) : [],
                    score: stroke.score || 0,
                    comment: stroke.comment || ''
                })),
                score: strokeData.score || 0,
                comment: strokeData.comment || ''
            };

            console.log('Request body:', requestBody);
            
            const response = await apiClient.post(`/api/characters/${characterId}/stroke-result`, requestBody);
            return response.data;
        } catch (error) {
            console.error('Error saving stroke result:', error);
            throw error.response?.data || error;
        }
    },

    // なぞり結果を取得
    async getStrokeResult(characterId) {
        try {
            const response = await apiClient.get(`/api/characters/${characterId}/stroke-result`);
            return response.data;
        } catch (error) {
            console.error('Error fetching stroke result:', error);
            throw error.response?.data || error;
        }
    },

    // すべてのなぞり結果を取得
    getAllStrokeResults: async (characterId) => {
        try {
            const response = await apiClient.get(`/api/characters/${characterId}/stroke-results`);
            return response.data;
        } catch (error) {
            console.error('Error fetching stroke results:', error);
            throw error;
        }
    }
}; 